sycl/doc/OptionalDeviceFeatures.md

```diff
@@ -547,12 +547,122 @@ example, if a new device is released before there is a new DPC++ release.  In
+the incoming monolythic device code module into pieces - smaller device code
+modules - it outputs a file table as a result. Each row in the table corresponds
+to an individual output module, and each element of a row is a name of a file
+containing necessary information about the module, such as the code itself, its
+properties.
+
+At the action graph building stage for each requested AOT compilation target -
+SPIR-V-based (such as Gen targets) and/or non-SPIR-V-based (such as PTX) - the
+driver adds an `aspect-filter` action which filters out input file table rows
+with device code modules using features unsupported on current target. Then the
+output table goes as input into the AOT stage, and the prior filtering
+guarantees that the AOT compiler will not encounter device code it can't
+compile. In the extreme case when all device code
+modules use unsupported aspects, the input file table will be empty. The picture
+below illustrates the action graph built by the clang driver along with file
+lists and tables generated and consumed by various nodes of the graph. The
+example set of targets used for the illustration is 4 targets
+- spir64 (runtime JITted SPIR-V)
+- AOT targets
+    - non-SPIR-V based
+        - nvptx (PTX)
+    - SPIR-V based
+        - gen_11 (Intel Gen)
+        - x86_64_avx512 (AVX512)
+<br>
+<br>
+
+![Device SPIRV translation and AOT compilation](images/DeviceLinkAOTAndWrap.svg)
+
+##### Aspect filter tool
+
+This tool transforms an input file table by removing rows with device code files
+that use features unsupported for the target architecture given as tool's
+argument.
+
+*Name*:
+- `sycl-aspect-filter`, located next to other tools like `file-table-tform`
+
+*Input*:
+- file table, normally coming out of `sycl-post-link` or `file-table-tform`
+  tools
+
+*Command line arguments*
+- `-target=<target>` target device architecture to filter for
+- `-device-config-file=<path>` path to the device configuration file
+
+*Output*
+- the input file table filtered as needed
+
+In more details, the tool performs the following actions:
+1) Checks if the input file table contains "Properties" column. If not, copies
+   the input file table to output and exits without error.
+1) Reads in the device configuration file and finds some entry `E` corresponding
+   to the architecture given on the command line. If there is no such entry -
+   reports and error and exits.
+1) For each row in the input file table:
+   - loads the properties file from the "Properties" column
+   - checks if there is the `SYCL/image-requirements` property
+   - if no, copies current row to the output file table and goes to the next
+   - if yes, checks if all the requirements listed in the property are supported
+     by the target architecture as specified by entry `E` in the device configuration file
+       - if yes, copies current row to the output file table and goes to the
+         next
+       - otherwise skips this row
+
```

Comment:
Let's say that for certain modules none of the requirements listed are supported for the target.  How will that be handled?  Perhaps we would error out as the desired files would not be produced.

Reply:
Sounds reasonable. It seems that if there are no suitable code modules for an AOT target appearing in the -fsycl-target list, then it shouldn't have appeared there. We can change later if encounter cases where this can happen for good reason.

Reply:
I think I do not agree, but let me make sure I understand the situation.  Are you suggesting that the line that says:

> otherwise skips this row

should be changed to:

> otherwise issue an error

That would definitely **not** be correct.  See the requirements section ["Diagnostics from the front-end compiler"](https://github.com/gmlueck/llvm/blob/gmlueck/optional-feature-design/sycl/doc/OptionalDeviceFeatures.md#diagnostics-from-the-front-end-compiler) which says:

> Note that this behavior does not change when the compiler runs in AOT mode. Even if the user specifies a target device via "-fsycl-targets", that does not necessarily mean that the user expects all the code in the application to be runnable on that device. Consider an application that uses some middleware library, where the library's header contains kernels optimized for different devices. An application should be able to compile in AOT mode with this library without getting errors. Therefore the AOT compiler must not fail simply because the middleware header contains device code for devices that are not being compiled for.

Reply:
We talk about a different situation here. When all the rows are skipped and the result of aspect-filter is empty file table. The logic is: this there is nothing this AOT target can compile, so specifying it on the -fsycl-targets list was a mistake.